File: week_1/day-5-string-hashing/first_non_repeating_char_follow_up_questions/first_non_repeating_stream.py
# Program: first_non_repeating_stream.py
from collections import defaultdict, deque

stream = "leetcode"
freq = defaultdict(int)
queue = deque()

for ch in stream:
    freq[ch] += 1
    queue.append(ch)

    while queue and freq[queue[0]] > 1:
        queue.popleft()

print(queue[0] if queue else -1)  # Output: l

File: week_1/day-5-string-hashing/first_non_repeating_char_follow_up_questions/firstNonRepeatingStream.js
// Program: firstNonRepeatingStream.js
const stream = "leetcode";
const freq = {};
const queue = [];

for (let ch of stream) {
    freq[ch] = (freq[ch] || 0) + 1;
    queue.push(ch);

    while (queue.length && freq[queue[0]] > 1) {
        queue.shift();
    }
}

console.log(queue.length ? queue[0] : -1); // l
